manubot/cite/citeproc.py

```diff
@@ -47,51 +14,133 @@
+        csl_item, = remove_jsonschema_errors([csl_item])
+
+    # Default CSL type to entry
+    csl_item['type'] = csl_item.get('type', 'entry')
+
+    if prune:
+        # Confirm that corrected CSL validates
+        validator = get_jsonschema_csl_validator()
+        validator.validate([csl_item])
+    return csl_item
+
+
+@functools.lru_cache()
+def get_jsonschema_csl_validator():
+    """
+    Return a jsonschema validator for the CSL Item JSON Schema
+    """
+    url = 'https://github.com/dhimmel/schema/raw/manubot/csl-data.json'
+    # Use jsonref to workaround https://github.com/Julian/jsonschema/issues/447
+    schema = jsonref.load_uri(url, jsonschema=True)
+    Validator = jsonschema.validators.validator_for(schema)
+    Validator.check_schema(schema)
+    return Validator(schema)
+
+
+def remove_jsonschema_errors(instance):
+    """
+    Remove fields in CSL Items that produce JSON Schema errors. Note that this
+    method may not be work for all types of JSON Schema errors and users
+    looking to adapt it for other applications should write task-specific tests
+    to provide empirical evaluate that it works as intended.
+
+    See also:
+    https://github.com/Julian/jsonschema/issues/448
+    https://stackoverflow.com/questions/44694835
+    """
+    validator = get_jsonschema_csl_validator()
+    errors = list(validator.iter_errors(instance))
+    instance = copy.deepcopy(instance)
+    errors = sorted(errors, key=lambda e: e.path, reverse=True)
+    for error in errors:
+        _remove_error(instance, error)
+    return instance
+
+
+def _delete_elem(instance, path, absolute_path=None, message=''):
+    """
+    Helper function for remove_jsonschema_errors that deletes an element in the
+    JSON-like input instance at the specified path. absolute_path is relative
+    to the original validated instance for logging purposes. Defaults to path,
+    if not specified. message is an optional string with additional error
+    information to log.
+    """
+    if absolute_path is None:
+        absolute_path = path
+    logging.debug(
+        (f'{message}\n' if message else message) +
+        '_delete_elem deleting CSL element at: ' +
+        '/'.join(map(str, absolute_path))
+    )
+    *head, tail = path
     try:
-        value = csl_item['issued']['date-parts'][0][0]
-        if value is None:
-            del csl_item['issued']
+        del _deep_get(instance, head)[tail]
     except KeyError:
         pass
 
-    return csl_item
+
+def _deep_get(instance, path):
+    """
+    Descend path to return a deep element in the JSON object instance.
+    """
+    for key in path:
+        instance = instance[key]
+    return instance
+
+
+def _remove_error(instance, error):
+    """
+    Remove a jsonschema ValidationError from the JSON-like instance.
+
+    See ValidationError documentation at
+    http://python-jsonschema.readthedocs.io/en/latest/errors/#jsonschema.exceptions.ValidationError
```

Comment:
It's somewhat difficult to follow this logic without reading the `ValidationError` documentation.  It's probably okay because you link the documentation for any future maintainers.

Reply:
Yeah, I myself don't even have a great understanding of the `jsonschema` package. It's possible there are types of errors that could occur that this function doesn't handle. However, with the current CSL Schema, I'm not sure they will come up.

Reply:
In the case of this `_remove_error` function, the fewer comments the better as there is a chance my comments would be wrong. In other words, to understand it, one should reference the code. I agree it's unfortunately complex.